rpc/blockchain.go

```diff
@@ -921,8 +872,162 @@ func (s *PublicBlockchainService) SetNodeToBackupMode(ctx context.Context, isBac
 	return s.hmy.NodeAPI.SetNodeBackupMode(isBackup), nil
 }
 
-func combineCacheKey(number uint64, version Version, blockArgs *rpc_common.BlockArgs) string {
-	// no need format blockArgs.Signers[] as a part of cache key
-	// because it's not input from rpc caller, it's caculate with blockArgs.WithSigners
-	return strconv.FormatUint(number, 10) + strconv.FormatInt(int64(version), 10) + strconv.FormatBool(blockArgs.WithSigners) + strconv.FormatBool(blockArgs.FullTx) + strconv.FormatBool(blockArgs.InclStaking)
+const (
+	blockCacheSize      = 2048
+	signersCacheSize    = blockCacheSize
+	stakingTxsCacheSize = blockCacheSize
+	leaderCacheSize     = blockCacheSize
+)
+
+type (
+	// bcServiceHelper is the getHelper for block factory. Implements
+	// rpc_common.BlockDataProvider
+	bcServiceHelper struct {
+		version Version
+		hmy     *hmy.Harmony
+		cache   *bcServiceCache
+	}
+
+	bcServiceCache struct {
+		signersCache    *lru.Cache // numberU64 -> []string
+		stakingTxsCache *lru.Cache // numberU64 -> interface{} (v1.StakingTransactions / v2.StakingTransactions)
+		leaderCache     *lru.Cache // numberUint64 -> string
+	}
+)
+
+func (s *PublicBlockchainService) newHelper() *bcServiceHelper {
+	signerCache, _ := lru.New(signersCacheSize)
+	stakingTxsCache, _ := lru.New(stakingTxsCacheSize)
+	leaderCache, _ := lru.New(leaderCacheSize)
```

Comment:
Understand, this is the previous cache the entire json body into three separate cache, according to the blockArgs judgment to fetch data separately？This is more elegant. and no need blockArg as a part of cacheKey